File: pgml-sdks/python/pgml/examples/question_answering_instructor.py
from pgml import Database
import os
from datasets import load_dataset
from time import time
from dotenv import load_dotenv
from rich.console import Console
import asyncio

async def main():
    load_dotenv()
    console = Console()

    local_pgml = "postgres://postgres@127.0.0.1:5433/pgml_development"

    conninfo = os.environ.get("PGML_CONNECTION", local_pgml)
    db = Database(conninfo)

    collection_name = "squad_collection"
    collection = await db.create_or_get_collection(collection_name)



    data = load_dataset("squad", split="train")
    data = data.to_pandas()
    data = data.drop_duplicates(subset=["context"])

    documents = [
        {"id": r["id"], "text": r["context"], "title": r["title"]}
        for r in data.to_dict(orient="records")
    ]

    console.print("Upserting documents ..")
    await collection.upsert_documents(documents[:200])
    console.print("Generating chunks ..")
    await collection.generate_chunks(splitter_id=1)

    # register instructor model
    console.print("Registering instructor model ..")
    instructor_model = "hkunlp/instructor-base" 
    instructor_model_params = {"instruction": "Represent the Wikipedia document for retrieval: "}

    model_id = await collection.register_model(
        model_name=instructor_model,
        model_params=instructor_model_params,
    )


    console.print("Generating embeddings .. for model %s" % (model_id), style="bold")
    await collection.generate_embeddings(model_id=model_id)

    console.print("Querying ..")
    start = time()
    query = "Who won 20 grammy awards?"
    results = await collection.vector_search(
        query,
        top_k=5,
        model_id=model_id,
        query_params={
            "instruction": "Represent the Wikipedia question for retrieving supporting documents: "
        },
    )
    _end = time()
    console.print("\nResults for '%s'" % (query), style="bold")
    console.print(results)
    console.print("Query time = %0.3f" % (_end - start))

    await db.archive_collection(collection_name)

if __name__ == "__main__":
    asyncio.run(main())

File: pgml-sdks/pgml/javascript/examples/question_answering_instructor.js
const pgml = require("pgml");
require("dotenv").config();

const main = async () => {
  // Initialize the collection
  const collection = pgml.newCollection("qa_pipeline");

  // Add a pipeline
  const pipeline = pgml.newPipeline("qa_pipeline", {
    text: {
      splitter: { model: "recursive_character" },
      semantic_search: {
        model: "Alibaba-NLP/gte-base-en-v1.5",
      },
    },
  });
  await collection.add_pipeline(pipeline);

  // Upsert documents, these documents are automatically split into chunks and embedded by our pipeline
  const documents = [
    {
      id: "Document One",
      text: "PostgresML is the best tool for machine learning applications!",
    },
    {
      id: "Document Two",
      text: "PostgresML is open source and available to everyone!",
    },
  ];
  await collection.upsert_documents(documents);

  // Perform vector search
  const query = "What is the best tool for building machine learning applications?";
  const queryResults = await collection.vector_search(
    {
      query: {
        fields: {
          text: {
            query: query,
            parameters: {
              instruction:
                "Represent the Wikipedia question for retrieving supporting documents: ",
            }
          }
        }
      }, limit: 1
    }, pipeline);
  console.log(queryResults);

  // Archive the collection
  await collection.archive();
};

main().then(() => console.log("Done!"));
